sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypes.scala

```diff
@@ -62,9 +62,14 @@ case class CreateStruct(children: Seq[NamedExpression]) extends Expression {
   override lazy val dataType: StructType = {
     assert(resolved,
       s"CreateStruct contains unresolvable children: ${children.filterNot(_.resolved)}.")
-    val fields = children.map { child =>
-      StructField(child.name, child.dataType, child.nullable, child.metadata)
-    }
+      val fields = children.zipWithIndex.map { case (child, idx) =>
+        child match {
+          case ne: NamedExpression =>
```

Comment:
The `CreateStruct` here was also used by `org.apache.spark.sql.functions.struct` and I did this above to avoid change the semantic of these two `struct` method. Maybe it's better to add another two class to do what `GenericUDFStruct ` & `GenericUDFNamedStruct` does?